networkx/algorithms/causal/m_separation.py

```diff
@@ -1,21 +1,35 @@
+from copy import deepcopy
+from collections import deque
+
 import networkx as nx
+from networkx.utils import UnionFind
+
+import graphs
 
 __all__ = ["m_separated"]
 
 
+def markov_blanket(
+    G, x, bidirected_edge_name="bidirected", directed_edge_name="directed"
+):
+    pass
+
+
 def m_separated(
     G, x, y, z, bidirected_edge_name="bidirected", directed_edge_name="directed"
 ):
     """Check m-separation among 'x' and 'y' given 'z' in mixed-edge causal graph G.
 
-    This algorithm wraps ``networkx.algorithms.d_separated``, but
-    allows one to pass in a ``ADMG`` instance instead.
+    This algorithm adapts the linear time algorithm presented in [1] currently implemeted in
+    networkx.algorithms.d_separation to work for mixed-edge causal graphs.
+
+    This algorithm first obtains the ancestral subgraph of x | y | z which only requires knowledge of
+    the directed edges. Then, all outgoing directed edges from nodes in z are deleted. After that, an
+    undirected graph composed from the directed and bidirected edges amongst the remaining nodes is
+    created. Then, x is independent of y given z if x is disconnected from y in this new graph.
 
-    It first converts all bidirected edges into explicit unobserved
-    confounding nodes in an explicit ``networkx.DiGraph``, which then
-    calls ``networkx.algorithms.d_separated`` to determine d-separation.
-    This inherently increases the runtime cost if there are many
-    bidirected edges, because many nodes must be added.
+    [1] Darwiche, A.  (2009).  Modeling and reasoning with Bayesian networks.
+       Cambridge: Cambridge University Press.
```

Comment:
In networkx, they cite things with the patterns seen in this function: https://github.com/networkx/networkx/blob/d2ceeb69ab0580a84a3b13266e6bd32e99da0232/networkx/algorithms/d_separation.py#L331-L333

Reply:
So a separate `References` section

Reply:
Yep copied the pattern